backend/tournesol/views/exports.py

```diff
@@ -35,6 +35,39 @@ def write_comparisons_file(request, write_target):
     )
 
 
+def write_public_comparisons_file(request, write_target):
+    """
+    Writes all public comparisons data as a CSV file to write_target which can be
+    among other options an HttpResponse or a StringIO
+    """
+    fieldnames = ['public_username', 'video_a', 'video_b', 'criteria', 'weight', 'score']
+    writer = csv.DictWriter(write_target, fieldnames=fieldnames)
+    writer.writeheader()
+    public_data = ContributorRating.objects.filter(is_public=True).select_related("user", "video")
+    serialized_comparisons = []
+    public_videos = set((rating.user, rating.video) for rating in public_data)
+    comparisons = Comparison.objects.all()\
+        .select_related("video_1", "video_2", "user")\
+        .prefetch_related("criteria_scores")
+    public_comparisons = [
+        comparison for comparison in comparisons
+        if ((comparison.user, comparison.video_1) in public_videos
+            and (comparison.user, comparison.video_2) in public_videos)]
+    public_usernames = [comparison.user.username for comparison in public_comparisons]
+    serialized_comparisons = ComparisonSerializer(public_comparisons, many=True).data
+    writer.writerows(
+        {
+            "public_username": public_username,
+            "video_a": comparison["video_a"]["video_id"],
+            "video_b": comparison["video_b"]["video_id"],
+            **criteria_score
+        }
+        for public_username in public_usernames
+        for comparison in serialized_comparisons
+        for criteria_score in comparison['criteria_scores']
```

Comment:
A bug slipped through here. The nested loop is incorrect: each comparison would be duplicated and appear with all usernames, instead of being associated with the correct username. Perhaps the test case could be extended to cover this situation (with multiple usernames in the public dataset).

A possible solution:
```suggestion
        for (public_username, comparison) in zip(public_usernames, serialized_comparisons)
        for criteria_score in comparison['criteria_scores']
```

Reply:
Resolved with commit e025b17